lms/services/canvas_studio.py

```diff
@@ -537,22 +539,46 @@ def _admin_oauth_http(self) -> OAuthHTTPService:
         # and use the standard `self._oauth_http_service` property instead.
         assert not self.is_admin()
 
+        # Find the (user_id, application_instance_id) for an admin user who
+        # belongs to the same LMS as the user making the request, and has
+        # authenticated with Canvas Studio.
+        #
+        # Note that the `application_instance_id` in this key can be different
+        # than that of the user making the request, since Hypothesis can be
+        # installed multiple times at one LMS, and the admin user may have
+        # authenticated using a different AI.
+        #
+        # If there are multiple matching `User` entities, pick the one that
+        # most recently authenticated with Canvas Studio.
+        guid = self._request.lti_user.application_instance.tool_consumer_instance_guid
         admin_email = self._admin_email()
         admin_user = (
             self._request.db.query(User)
-            .filter_by(
-                email=admin_email, application_instance=self._application_instance
+            .join(ApplicationInstance)
+            .join(
+                OAuth2Token,
+                and_(
+                    OAuth2Token.user_id == User.user_id,
+                    OAuth2Token.application_instance_id == ApplicationInstance.id,
+                ),
             )
-            .one_or_none()
+            .filter(User.email == admin_email)
+            .filter(ApplicationInstance.tool_consumer_instance_guid == guid)
+            .filter(OAuth2Token.service == Service.CANVAS_STUDIO)
+            .order_by(OAuth2Token.received_at.desc())
+            .first()
```

Comment:
An issue I noticed when `EXPLAIN`-ing this query locally is that there is no index on `User.email`. As a result this has to do a sequential scan of the `user` table. This is already the case with the existing query, but something we should aim to fix.

Reply:
For future reference we added an index on `ApplicationInstance.tool_consumer_instance_guid`

https://github.com/hypothesis/lms/pull/6410